### PR TITLE
tests: drivers: counter: basic_api: Fix for failures on nrf54h20/cpuppr

### LINKED ISSUE #75450: tests: drivers: counter: counter_basic_api: drivers.counter.basic_api.nrf54h20_cpuppr fails on nrf54h20 ppr
**Describe the bug**
The tests:
-tests/drivers/counter/counter_basic_api/drivers.counter.basic_api.nrf54h20_cpuppr

Observed for
- nrf54h20dk/nrf54h20/cpuppr

**To Reproduce**
1. have nrf54h20dk/nrf54h20/cpuppr connected
2. go to your zephyr dir
3. call ./scripts/twister -T tests/drivers/counter/counter_basic_api -p nrf54h20dk/nrf54h20/cpuppr --device-testing --device-serial /dev/ttyACM1 -v --inline-logs
4. See console output with error

**Expected behavior**
Valid console output

**Impact**
Not clear

**Logs and console output**
tests/kernel/interrupt/arch.interrupt.minimallibc:
```
 ERROR   - *** Booting Zephyr OS build v3.7.0-rc2-61-g6700a154be3c ***
Running TESTSUITE counter_basic
===================================================================
START - test_all_channels
Testing timer@9a2000
Testing timer@9a3000
Testing timer@9b2000
Testing timer@9b3000
Testing timer@9c2000
Testing timer@9c3000
Testing timer@9d2000
Testing timer@9d3000
Testing rtc@928000
Testing rtc@929000
PASS - test_all_channels in 1.331 seconds
===================================================================
START - test_cancelled_alarm_does_not_expire
Testing timer@9a2000
Testing timer@9a3000
Testing timer@9b2000
Testing timer@9b3000
Testing timer@9c2000
Testing timer@9c3000
Testing timer@9d2000
Testing timer@9d3000
Assertion failed at ../../../../../../../../../../zephyr/zephyr/tests/drivers/counter/counter_basic_api/src/test_counter.c:951: test_cancelled_alarm_does_not_expire_instance: (0 not equal to cnt)
timer@9d3000: Expected 0 callbacks, got 1 (i:744)
FAIL - test_cancelled_alarm_does_not_expire in 16.807 seconds
===================================================================
START - test_late_alarm
Testing timer@9a2000
Testing timer@9a3000
Testing timer@9b2000
Testing timer@9b3000
Testing timer@9c2000
Testing timer@9c3000
Testing timer@9d2000
Testing timer@9d3000
Assertion failed at ../../../../../../../../../../zephyr/zephyr/tests/drivers/counter/counter_basic_api/src/test_counter.c:703: test_late_alarm_instance: (-ETIME not equal to err)
timer@9d3000: Unexpected error (0)
FAIL - test_late_alarm in 0.745 seconds
===================================================================
START - test_late_alarm_error
Testing timer@9a2000
Testing timer@9a3000
Testing timer@9b2000
Testing timer@9b3000
Testing timer@9c2000
Testing timer@9c3000
Testing timer@9d2000
Testing timer@9d3000
Assertion failed at ../../../../../../../../../../zephyr/zephyr/tests/drivers/counter/counter_basic_api/src/test_counter.c:754: test_late_alarm_error_instance: (-ETIME not equal to err)
timer@9d3000: Failed to detect late setting (err: -16)
FAIL - test_late_alarm_error in 0.748 seconds
===================================================================
START - test_multiple_alarms
Testing timer@9a2000
Testing timer@9a3000
Testing timer@9b2000
Testing timer@9b3000
Testing timer@9c2000
Testing timer@9c3000
Testing timer@9d2000
Testing timer@9d3000
Assertion failed at ../../../../../../../../../../zephyr/zephyr/tests/drivers/counter/counter_basic_api/src/test_counter.c:565: test_multiple_alarms_instance: (0 not equal to err)
timer@9d3000: Counter failed to set top value
FAIL - test_multiple_alarms in 1.125 seconds
===================================================================
START - test_set_top_value_with_alarm
Testing timer@9a2000
Testing timer@9a3000
Testing timer@9b2000
Testing timer@9b3000
Testing timer@9c2000
Testing timer@9c3000
Testing timer@9d2000
Testing timer@9d3000
Assertion failed at ../../../../../../../../../../zephyr/zephyr/tests/drivers/counter/counter_basic_api/src/test_counter.c:277: test_set_top_value_with_alarm_instance: (0 not equal to err)
timer@9d3000: Counter failed to set top value (err: -16)
FAIL - test_set_top_value_with_alarm in 1.516 seconds
===================================================================
START - test_short_relative_alarm
Testing timer@9a2000
Testing timer@9a3000
Testing timer@9b2000
Testing timer@9b3000
Testing timer@9c2000
Testing timer@9c3000
Skipped for timer@9d2000
Skipped for timer@9d3000
Assertion failed at ../../../../../../../../../../zephyr/zephyr/tests/drivers/counter/counter_basic_api/src/test_counter.c:175: counter_tear_down_instance: ((err == 0) || (err == -ENOTSUP) is false)
timer@9d3000: Setting top value to default failed
FAIL - test_short_relative_alarm in 0.812 seconds
===================================================================
START - test_single_shot_alarm_notop
Testing timer@9a2000
Testing timer@9a3000
Testing timer@9b2000
Testing timer@9b3000
Testing timer@9c2000
Testing timer@9c3000
Testing timer@9d2000
Testing timer@9d3000
Assertion failed at ../../../../../../../../../../zephyr/zephyr/tests/drivers/counter/counter_basic_api/src/test_counter.c:433: test_single_shot_alarm_instance: (0 not equal to err)
timer@9d3000: Counter set alarm failed (err: -16)
FAIL - test_single_shot_alarm_notop in 1.237 seconds
===================================================================
START - test_single_shot_alarm_top
Testing timer@9a2000
Testing timer@9a3000
Testing timer@9b2000
Testing timer@9b3000
Testing timer@9c2000
Testing timer@9c3000
Testing timer@9d2000
Assertion failed at ../../../../../../../../../../zephyr/zephyr/tests/drivers/counter/counter_basic_api/src/test_counter.c:440: test_single_shot_alarm_instance: (1 not equal to cnt)
timer@9d2000: Expecting alarm callback
FAIL - test_single_shot_alarm_top in 1.103 seconds
===================================================================
TESTSUITE counter_basic failed.
Running TESTSUITE counter_no_callback
===================================================================
START - test_set_top_value_without_alarm
Testing timer@9a2000
Testing timer@9a3000
Testing timer@9b2000
Testing timer@9b3000
Testing timer@9c2000
Testing timer@9c3000
Testing timer@9d2000
Testing timer@9d3000
Assertion failed at ../../../../../../../../../../zephyr/zephyr/tests/drivers/counter/counter_basic_api/src/test_counter.c:326: test_set_top_value_without_alarm_instance: (0 not equal to err)
timer@9d3000: Counter failed to set top value (err: -16)
FAIL - test_set_top_value_without_alarm in 0.788 seconds
===================================================================
TESTSUITE counter_no_callback failed.
------ TESTSUITE SUMMARY START ------
SUITE FAIL -  11.11% [counter_basic]: pass = 1, fail = 8, skip = 0, total = 9 duration = 25.424 seconds
- PASS - [counter_basic.test_all_channels] duration = 1.331 seconds
- FAIL - [counter_basic.test_cancelled_alarm_does_not_expire] duration = 16.807 seconds
- FAIL - [counter_basic.test_late_alarm] duration = 0.745 seconds
- FAIL - [counter_basic.test_late_alarm_error] duration = 0.748 seconds
- FAIL - [counter_basic.test_multiple_alarms] duration = 1.125 seconds
- FAIL - [counter_basic.test_set_top_value_with_alarm] duration = 1.516 seconds
- FAIL - [counter_basic.test_short_relative_alarm] duration = 0.812 seconds
- FAIL - [counter_basic.test_single_shot_alarm_notop] duration = 1.237 seconds
- FAIL - [counter_basic.test_single_shot_alarm_top] duration = 1.103 seconds
SUITE FAIL -   0.00% [counter_no_callback]: pass = 0, fail = 1, skip = 0, total = 1 duration = 0.788 seconds
- FAIL - [counter_no_callback.test_set_top_value_without_alarm] duration = 0.788 seconds
------ TESTSUITE SUMMARY END ------
===================================================================
PROJECT EXECUTION FAILED
```

**Environment (please complete the following information):**
 - OS: Ubuntu 20.04.1 LTS
 - Toolchain Zephyr SDK 0.16.5
 - Commit SHA or Version used: v3.7.0-rc2-61-g6700a154be3c

### PR DESCRIPTION
**Pull request by nordic-krch**
Previous attempt (331442f05) was not correct and it only prolonged the test twice. It turned out that there was a bug in the driver. Reverting previous change and fixing the driver.

Fixes #75450.

### CODE REVIEW

File: drivers/counter/counter_nrfx_timer.c [hunk 1]
```diff
@@ -89,6 +90,7 @@ static uint32_t read(const struct device *dev)
 
 	nrf_timer_task_trigger(timer,
 			       nrf_timer_capture_task_get(COUNTER_READ_CC));
+	barrier_dmem_fence_full();
```
Comment (gmarull) [inline drivers/counter/counter_nrfx_timer.c:93]: shouldn't this be using `nrf_timer_task_trigger` and `nrfy_timer_task_trigger`? Isn't this the whole idea of nrfy?

Comment (nordic-krch) [reply drivers/counter/counter_nrfx_timer.c]: nrfy was created with assumption that any periperhal access from vpr core can be reordered but this is not the case anymore. Reordering can only occur between different type of operations (read or write). In this driver i've identified 2 places.
When i tried to convert driver to use nrfy_timer then it gets +200 bytes of code. Given that vpr cores have very limited memory imo it's better to have barriers only where necessary.

Comment (gmarull) [reply drivers/counter/counter_nrfx_timer.c]: so, is gonna nrfy be removed?

**Review by gmarull — COMMENTED** [review]
I think in drivers, we should be using the Zephyr barrier API

**Comment by nordic-krch** [comment]
I used nrf api because barrier is needed only for VPR core.

File: drivers/counter/counter_nrfx_timer.c (hunk 1 above)
Comment (nordic-krch) [reply drivers/counter/counter_nrfx_timer.c]: we will discuss it internally but seems that there is no point of using barriers everywhere if they are needed only in very limited number of places.

File: drivers/counter/counter_nrfx_timer.c
[the remark this answers is not in the document]
Comment (nordic-krch) [reply drivers/counter/counter_nrfx_timer.c:1]: see comment above https://github.com/zephyrproject-rtos/zephyr/pull/76955#discussion_r1714849151

[the remark this answers is not in the document]
Comment (kl-cruz) [reply drivers/counter/counter_nrfx_timer.c:1]: I approve it to move forward with correct solution. We will discuss it internally

**Review by kl-cruz — APPROVED** [review]

**Review by masz-nordic — APPROVED** [review]

**Pull request merged**